### PR TITLE
Add a Security Model doc

### PR DESCRIPTION
**Pull request by lucacasonato**
- Add a Security Model doc

### CODE REVIEW

File: runtime/reference/security_model.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
+---
+title: "Security Model"
+---
+
+This document outlines the security model of the Deno runtime. Deno is designed
+to be secure by default runtime, however to write secure applications it is
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:6]:
```suggestion
to be secure-by-default runtime, however to write secure applications it is
```
does this make sense? I read this sentence three 🤣

File: runtime/reference/security_model.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  spawn subprocesses.
+- **No limits on the execution of code at the same privilege level**: Deno
+  allows the execution of any code (JS/TS/Wasm) via multiple means, including
+  `eval`, `new Function`, dynamic imports and web workers at the same privilege
+  level with little restriction as to where the code originates (network, npm,
+  JSR, etc).
+- **Multiple invocations of the same application can share data**: Deno provides
+  a mechanism for multiple invocations of the same application to share data,
+  through built in caching and KV storage APIs. Different applications can not
+  see each other's data.
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:23]: Should we mention here that it's still possible with correct combination of flags to see this data?

File: runtime/reference/security_model.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+- **All code executing on the same thread shares the same privilege level**: All
+  code executing on the same thread shares the same privilege level. It is not
+  possible for different modules to have different privilege levels within the
+  same thread.
+- **Code can not escalate it's privileges without user consent**: Code executing
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:28]:
```suggestion
- **Code can not escalate its privileges without user consent**: Code executing
```

File: runtime/reference/security_model.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  in a Deno runtime can not escalate it's privileges without the user agreeing
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:29]:
```suggestion
  in a Deno runtime can not escalate its privileges without the user agreeing
```

File: runtime/reference/security_model.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  explicitly to an escalation.
+- **The initial static module graph can import local files without
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:31]: What about remote modules?

File: runtime/reference/security_model.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+## Permissions
+
+By default, access to most system I/O is denied. There are some I/O operations
+that are allowed in a limited capacity, even by default. These are described
+below.
+
+To enable these operations, the user must explicitly grant permission to the
+Deno runtime. This is done by passing the `--allow-read`, `--allow-write`,
+`--allow-net`, `--allow-env`, and `--allow-run` flags to the `deno` command.
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:51]: I was about to suggest you link to https://docs.deno.com/runtime/fundamentals/security/ but maybe we should roll this document into that already existing document?

File: runtime/reference/security_model.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+- All files that are imported from the entrypoint module in a way that they can
+  be statically analyzed are allowed to be read by default. This includes static
+  `import` statements and dynamic `import()` calls where the argument is a
+  string literal that points to a specific file or a directory of files. The
+  full list of files that are in this list can be printed using
+  `deno info <entrypoint>`.
+- Files that are dynamically imported in a way that can not be statically
+  analyzed require runtime read permissions.
+- Files inside of a `node_modules/` directory are allowed to be read by default.
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:89]: But not if used with `require()`, let's mention this here

File: runtime/reference/security_model.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+- `https://gist.githubusercontent.com`
+
+These locations are trusted "public good" registries that are not expected to
+enable data exfiltration through URL paths.
+
+In addition Deno allows importing any NPM package through `npm:` specifiers.
+
+Deno also sends requests to `https://dl.deno.land/` at most once a day to check
+for updates to the Deno CLI. This can be disabled using
+`DENO_NO_UPDATE_CHECK=1`.
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:119]:
```suggestion
`DENO_NO_UPDATE_CHECK=1` environmental variable.
```

File: runtime/reference/security_model.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+### Environment variables
+
+By default, executing code can not read or write environment variables. This
+includes reading environment variables, and setting new values.
+
+Deno reads certain environment variables on startup, such as `DENO_DIR` and
+`NO_COLOR` (see `deno help` for the full list).
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:127]: Let's link to https://docs.deno.com/runtime/reference/cli/env_variables/ instead

File: runtime/reference/security_model.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+The value of the `NO_COLOR` environment variable is visible to all code running
+in the Deno runtime, regardless of whether the code has been granted permission
+to read environment variables.
+
+### Subprocesses
+
+Code executing inside of a Deno runtime can not spawn subprocesses by default,
+as this would constitute a violation of the principle that code can not escalate
+it's privileges without user consent.
```
Comment (bartlomieju) [inline runtime/reference/security_model.md:137]:
```suggestion
its privileges without user consent.
```

File: runtime/reference/security_model.md (hunk 6 above)
Comment (thisisjofrank) [reply runtime/reference/security_model.md:51]: seconded

File: runtime/reference/security_model.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+Deno provides a mechanism for executing code written in other languages, such as
+Rust, C, or C++, from within a Deno runtime. This is done using the
+`Deno.dlopen` API, which can load shared libraries and call functions from them.
+
+By default, executing code can not use the `Deno.dlopen` API, as this would
+constitute a violation of the principle that code can not escalate it's
+privileges without user consent.
+
+In addition to `Deno.dlopen`, FFI can also be used via Node-API (NAPI) native
+addons. These are also not allowed by default.
```
Comment (littledivy) [inline runtime/reference/security_model.md:157]: I think this should include "how to allow them" via the flag. Slightly worried that search engine summarization will assume that it's just not allowed

File: runtime/reference/security_model.md (hunk 2 above)
Comment (lucacasonato) [reply runtime/reference/security_model.md:23]: I think it is broken right now.

File: runtime/reference/security_model.md (hunk 7 above)
Comment (lucacasonato) [reply runtime/reference/security_model.md:89]: wdym?

File: runtime/reference/security_model.md (hunk 5 above)
Comment (lucacasonato) [reply runtime/reference/security_model.md:31]: They can import local files via data urls: https://gist.githubusercontent.com/lucacasonato/68811668a46849cc6201f5bb9f2d4185/raw/e7bbe26bd6d120d2b6091a67043d217e00c1b428/test.ts

File: runtime/reference/security_model.md (hunk 7 above)
Comment (bartlomieju) [reply runtime/reference/security_model.md:89]: If you have `require()` calls in your code you will still need to add `--allow-read` flag

**Review by bartlomieju — APPROVED** [review]

File: runtime/reference/security_model.md (hunk 7 above)
Comment (lucacasonato) [reply runtime/reference/security_model.md]: Yeah, but not to read the `node_modules/` folder. You need it only to read the `package.json` in CWD. You can do this completely without any permissions for example:

```js
import { createRequire } from "node:module";

// import require from "c";
const require = createRequire(
  new URL("./node_modules/chalk", import.meta.url).href,
);

const chalk = require("chalk");
console.log(chalk.blue("Hello world!"));
```

**Pull request merged**